File: C#/4_Merge Sort/Merge Sort.cs
using System;

namespace Algorithm {

    class MergeSort {

        public static void merge(int[] arr, int l, int m, int r){
            int n1 = m - l + 1;
            int n2 = r - m;

            int[] L = new int[n1];
            int[] R = new int[n2];

            for(int x = 0; x < n1; x++){
                L[x] = arr[l+x];
            }

            for(int y = 0; y < n2; y++){
                R[y] = arr[m+1+y];
            }

            int i = 0, j = 0;
            int k = l;

            while(i < n1 && j < n2){
                if(L[i] <= R[j]){
                    arr[k] = L[i];
                    i++;
                }else{
                    arr[k] = R[j];
                    j++;
                }
                k++;
            }

            while(i < n1){
                arr[k] = L[i];
                i++;
                k++;
            }

            while(j < n2){
                arr[k] = R[j];
                j++;
                k++;
            }
        }

        public static void mergeSort(int[] arr, int l, int r){
            if(l < r) {
                int m = (l + r) / 2;
                mergeSort(arr, l, m);
                mergeSort(arr, m + 1, r);
                merge(arr, l, m, r);
            }
        }

        public static void printArray(int[] arr){
            for(int i = 0; i < arr.Length; i++){
                Console.Write(arr[i] + " ");
            }
        }

        public static void Main(string[] args) {
            int[] arr = {10, 54, 63, 6, 420};
            mergeSort(arr, 0, arr.Length - 1);
            printArray(arr);
        }
    }
}

File: C/4_Merge Sort/Merge Sort.c
#include <stdio.h>
#include <stdlib.h>

// Prototipi di funzioni
void merge(int arr[], int from , int to, int mid);
void mergeSort(int arr[], int from, int to);
void printArray(int arr[], int n);

void merge(int arr[], int from , int to, int mid){
    int i = from;
    int j = mid + 1;
    int k = 0;

    int temp[to-from+1];

    while(i <= mid && j <= to){
        if(arr[i] < arr[j]){
            temp[k] = arr[i];
            k++;
            i++;
        }else{
            temp[k] = arr[j];
            k++;
            j++;
        }
    }

    while(i <= mid){
        temp[k] = arr[i];
        k++;
        i++;
    }

    while(j <= to){
        temp[k] = arr[i];
        k++;
        j++;
    }

    for(i = from; i <= to; i++){
        arr[i] = temp[i-from];
    }
}

void mergeSort(int arr[], int from , int to){
    int mid;

    if(from < to){
        mid = (from + to) / 2;
        
        mergeSort(arr, from, mid);
        mergeSort(arr, mid + 1, to);
        merge(arr, from, to, mid);
    }
}

void printArray(int arr[], int n){
    int i;
    for(i = 0; i < n; i++){
        printf("%d ", arr[i]);
    }
}

int main(int argc, char const *argv[]) {
    int arr[] = {10, 54, 63, 6, 40, 11, 41, 32, 17, 22};
	int n = sizeof(arr) / sizeof(arr[0]);
    mergeSort(arr, 0, n - 1);
    printArray(arr, n);
    return 0;
}
